fix: not handling array values in `aud` claim correctly (#4043)
PostgREST/postgrest: docs/references/auth.rst

- ``exp`` Expiration Time
- ``iat`` Issued At
- ``nbf`` Not Before
- ``aud`` :ref:`Audience <jwt_aud_validation>`

.. note::
  PostgREST allows for a 30-second clock skew when validating the ``exp``, ``iat`` and ``nbf`` claims.
  In other words, it gives an extra 30 seconds before the token is rejected if there is a slight discrepancy in the timestamps.

.. _jwt_aud_validation:

JWT ``aud`` Claim Validation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> PostgREST has a built in validation of the `JWT audience claim <https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.3>`_.
> It works this way:

Reviewer: ```suggestion
PostgREST has built in validation for the `JWT audience claim <https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.3>`_. It works this way:
```

> - By default, PostgREST identifies with all audiences and allows the JWT for any ``aud`` claim.
> - To change this, you can set a specific audience in the :ref:`jwt-aud` configuration.
>   PostgREST will then check if this audience is present in the ``aud`` claim list (or if it's the same value if ``aud`` is a string).
>   If this fails, then it rejects the token and responds with a :ref:`401 Unauthorized <pgrst303>` error.
> - Note that the token will not be rejected if the ``aud`` claim **is not present** or if its value is ``null`` or ``[]``.
>   PostgREST will interpret this token as allowed for all audiences and will complete the request.

Reviewer: @laurenceisla This seems more complete and understandable to me. WDYT?

```suggestion
- If :ref:`jwt-aud` is not set (the default), PostgREST identifies with all audiences and allows the JWT for any ``aud`` claim.
- If :ref:`jwt-aud` is set to a specific audience, PostgREST will check if this audience is present in the ``aud`` claim.
  + If the ``aud`` value is a JSON string, it will match it to the set :ref:`jwt-aud`.
  + If the ``aud`` value is a JSON array, it will search every element for a match.
  + If a match is not found or if above the ``aud`` value is a JSON object or number, then the token will be rejected with a :ref:`401 Unauthorized <pgrst303>` error.
  + If the ``aud`` key **is not present** or if its value is ``null`` or ``[]``, PostgREST will interpret this token as allowed for all audiences and will complete the request.
```

Author: > ...or if the ``aud`` value is a JSON object or number, then the token will be rejected ...

Instead of listing invalid values, maybe just mention that "if it's not an string or array of strings, then it will return an error". Or just "if it's an invalid value, then...".

Reviewer: First one looks good. Also we should add that the JSON array is composed of strings. I forgot that above.
